### ISSUE TITLE
Document CSS customization to the userbar (such as changing the icon)

### ISSUE THREAD

**Issue by Luperi** [wagtail/wagtail#2353]
Is there an easy way to customize userbar logo like the admin area one (as explained [here](http://docs.wagtail.io/en/v1.3.1/advanced_topics/customisation/branding.html#branding-logo))?

**Comment by gasman** [comment]
Not currently, no.

**Comment by alexgleason** [comment]
See #2373 for more on this. If you're using an icon font like FontAwesome you can just override the CSS to change the icon.

**Comment by alexgleason** [comment]
I don't think this is something that makes sense to implement. The icon is controlled via CSS and not an image, so overriding it is trivial.

We may want to document the process instead of implementing any kind of hook or override.

**Comment by SalahAdDin** [comment]
:+1:

**Comment by Luperi** [comment]
@alexgleason ok, good. So, should I have to use something like FontAwesome?

**Comment by gasman** [comment]
I haven't tried it, but I'd expect that setting a background image would work too. (If you did that, you'd probably also need something like `.wagtail-icon:before {content: ''}` to disable the existing icon.)

**Comment by alexgleason** [comment]
@Luperi Here's an example of one I made: https://www.youtube.com/watch?v=k2YPEUFGOgw

I think you might want to wait on #2373 before depending on this method. But yes, you can just set:

```css
.wagtail-icon:before {
    content: /* the CSS code for your FontAwesome icon/*;
    font-family: FontAwesome;
}
```

**Comment by inyoka** [comment]
You can recreate templates/wagtailadmin/userbar/base.html, here is an example using a fontawesome pencil icon as a replacement icon in the userbar ...

`{% load staticfiles i18n %}

<div class="wagtail-userbar-reset">
    <div class="wagtail-userbar wagtail-userbar--{{ position|default:'bottom-right' }}" data-wagtail-userbar>
        <link rel="stylesheet" href="{% static 'wagtailadmin/css/userbar.css' %}" type="text/css" />
        <div class="wagtail-userbar-nav">
            <div class="wagtail-userbar-trigger" data-wagtail-userbar-trigger>
              <i class="fas fa-pencil-alt fa-lg"></i>
                <span class="wagtail-userbar-help-text">{% trans 'My admin interface' %}</span>
            </div>
            <div class='wagtail-userbar-items'>
                {% for item in items %}
                    {{ item|safe }}
                {% endfor %}
            </div>
        </div>
        <script src="{% static 'wagtailadmin/js/userbar.js' %}"></script>
    </div>
</div>`

**Comment by Meteor0id** [comment]
I see @inyoka had the same idea I had. I noticed this issue report after writing the pull request.
If it does getmerged, you should soon be able to overwrite that wagtail branding logo with an image as well, just like you would with the branding_logo in other places on wagtail admin.

if it gets merged, you can create a file `dashboard/templates/wagtailadmin/userbar/base.html`:
```
{% extends "wagtailadmin/userbar/base.html" %}
{% load staticfiles %}

{% block branding_logo %}
    <img src="{% static 'ffs_base/img/my_logo.svg' %}" alt="my logo" />
{% endblock %}
```

**Comment by Meteor0id** [comment]
Note that I have not yet provided a way of customizing the background color (that bluish glow effect around the button). There is however another issue open discussing ways to customize wagtails color scheme. The background color for the userbar should be part of that.

The shape of the userbar button can also quite easily be manipulated in css / scss. However there first needs to be a consensus on a mechanism to allow developers to customize the scss. For now you would have to manually modify the css files (the class wagtail-userbar-trigger defined in userbar.css)

**Comment by Meteor0id** [comment]
I disagree with @alexgleason. It does make sense to implement. Some users might want to provide their icon or logo as an image, not as a font. Using the same methos as for the branding_logo on the admin page makes perfect sense.

Provided PR introduces that option, and the option to easily use a font, and the documentation on how to do so. Just waiting for it to get reviewd.

**Comment by thibaudcolas** [comment]
In my opinion, it would be better if `branding_logo` was replaced with a setting that only allows providing the logo as an image's URL/path rather than customising the HTML (and potentially CSS).

Then, it will be possible to:

- Only set the custom logo once for all of the places where logos are used (admin menu, userbar, 404 page)
- Always render logos as square, reducing the risk for issues like #4747

This of course limits how creative you can get with custom logos, but the constraints make it possible to ensure logos will be displayed consistently in any current and future UI – and it'll also make it easier to validate the resulting UI is accessible.

**Comment by gasman** [comment]
Given that the default Wagtail logo has custom JS behaviour of its own, it seems a bit unfair to deny that option to custom logos...

I think it would make sense to offer both options: have a setting for an image URL which overrides the Wagtail logo if specified, but keep them in `branding_logo` blocks that can be overridden for fancier customisations. Of course, at that point it becomes more acceptable to say "if you make non-trivial customisations to the logo using block overriding, it's your own responsibility to fix the CSS"...

**Comment by thibaudcolas** [comment]
Here's an example of the kind of issue the current implementation causes. The mobile nav has a different width when the page explorer is open, and custom icons end up having the "Dashboard" link displayed differently based on that:

![logo-customisation-2353](https://user-images.githubusercontent.com/877585/47614837-45387b80-dab7-11e8-9898-ce8e23db02fb.gif)

The default logo doesn't have this issue, since it has its own CSS to handle that case.

**Comment by gasman** [comment]
@thibaudcolas It's possible that people are already making non-trivial customisations (i.e. ones that go beyond setting an image) by overriding the `branding_logo`, and they'll be unhappy about us removing existing functionality. Is there any reason we can't keep the `branding_logo` block(s) in place, but make it clear in the docs that setting an image URL is the preferred route for simple logo customisations, and if you _do_ override the block, it's your own responsibility not to screw up the CSS?

**Comment by thibaudcolas** [comment]
Looking at #4731, it works as expected, and even though I’m not sold on the idea it seems to be solving a genuine problem for others – so this will now be the documented way of customizing the userbar’s logo from Wagtail 2.10 onwards:

In a `templates/wagtailadmin/userbar/base.html`:

```jinja
{% extends "wagtailadmin/userbar/base.html" %}

{% block branding_logo %}
    <img src="{% static 'app_name/img/my_logo.svg' %}" alt="" />
{% endblock %}
```

I would recommend people who rely on customizing this to migrate to this documented way of customizing the icon. While the CSS way might be easier currently, it’s undocumented and relies on implementation details that will go away once we merge #4821.
